### ISSUE TITLE
mAP drops to 0 when doing Knowledge Distillation with a detection network

### ISSUE THREAD

**Issue by RizhaoCai** [NervanaSystems/distiller#347]
Hello,

The Distiller is awesome and I would like to use it to prune my RFB-NET for detection.

There are two models, a RFB-NET with VGG-16 as its backbone (V) and another RFB-NET with as its  backbone (M).

The V could get 74 mAP while the M could only get 65 mAP in my case. I would like to get higer mAP with the M model and so I tried to use the Knowledge Distillation (KD) with the V as a teacher to guide the student M. 

Firstly, I used the KD to train a M model from scratch. However, after many epochs, I could get no more than 1 mAP. I thought there would be something wrong.

Then I loaded a pre-train M model, which can get 65 mAP and run the validation (test) before the training.
![image](https://user-images.githubusercontent.com/37952703/62674887-ff5dfb80-b9d6-11e9-98d1-0c96b1ab6c2a.png)

And I got  65.02 mAP
![image](https://user-images.githubusercontent.com/37952703/62674898-07b63680-b9d7-11e9-926d-968e6e74e436.png)

However, after one epoch of training and doing the test again, the mAP was 0
![image](https://user-images.githubusercontent.com/37952703/62674930-2d434000-b9d7-11e9-9771-3b0363a60b5c.png)

I have no idea how what is the problem.

------------------------------------------------------------------
Here is some code from the main script
```
# Loading models
model= build_net(mparams, num_classes=NUM_CLASSES).cuda()
model.load_state_dict(torch.load('./model/RFB_M.pth'))
teacher = Teacher(mparams, num_classes=NUM_CLASSES).cuda()
teacher.load_state_dict(torch.load('./models/RFB_VGG16.pth'))
teacher.eval()
# ...

args.kd_policy = None
global teacher

compression_scheduler = distiller.CompressionScheduler(model)
activations_collectors = create_activation_stats_collectors(model, *args.activation_stats)

 
kd_distill_wt, kd_student_wt, kd_teacher_wt, kd_temp = 0.5, 0.5, 0.0, 1
kd_start_epoch = 0
dlw = distiller.DistillationLossWeights(kd_distill_wt, kd_student_wt, kd_teacher_wt) # 
args.kd_policy = distiller.KnowledgeDistillationPolicy(model, teacher, kd_temp, dlw)
compression_scheduler.add_policy(args.kd_policy, starting_epoch=kd_start_epoch, ending_epoch=args.epochs, frequency=1)

for epoch in range(start_epoch, ending_epoch):
        # This is the main training loop.
        msglogger.info('\n')       
        if compression_scheduler:
            print('Scheduler: on epoch begin')
            compression_scheduler.on_epoch_begin(epoch)
        
        # Train for one epoch
        with collectors_context(activations_collectors["train"]) as collectors:
            print("Training")
            train(train_loader, model, anchors, criterion, optimizer, epoch, compression_scheduler,
                  loggers=[tflogger, pylogger], args=args)
            distiller.log_weights_sparsity(model, epoch, loggers=[tflogger, pylogger])
            distiller.log_activation_statsitics(epoch, "train", loggers=[tflogger],
                                                collector=collectors["sparsity"])
    
            if args.masks_sparsity:
                msglogger.info(distiller.masks_sparsity_tbl_summary(model, compression_scheduler))
                
        with collectors_context(activations_collectors["valid"]) as collectors:
            with torch.no_grad():
                mAP = test(testset, model, criterion, [pylogger], args, epoch)
            distiller.log_activation_statsitics(epoch, "valid", loggers=[tflogger],
                                                collector=collectors["sparsity"])
            save_collectors_data(collectors, msglogger.logdir)
            
        vloss = 0 # Validation Loss
        
        stats = ('Performance/Validation/',
                 #OrderedDict([('Loss', vloss),('val_mAP', mAP)] ) )
                 OrderedDict([('val_mAP', mAP)] ) 
                 )
        distiller.log_training_progress(stats, None, epoch, steps_completed=0, total_steps=1, log_freq=1,
                                        loggers=[tflogger])

        if compression_scheduler:
            #compression_scheduler.on_epoch_end(epoch, optimizer, metrics={'min': vloss, 'max': map})
            print('Scheduler: on epoch begin')
            compression_scheduler.on_epoch_end(epoch, optimizer, metrics={ 'max': mAP})

        # Update the list of top scores achieved so far, and save the checkpoint
        is_best = False
        ##global BEST_MAP, BEST_EPOCH
        if mAP > BEST_MAP:
            BEST_MAP = mAP
            BEST_EPOCH = epoch
            is_best = True
            
        #update_training_scores_history(perf_scores_history, model, top1, top5, epoch, args.num_best_scores)
        
        checkpoint_extras = {'current_mAP': mAP,
                             'best_mAP': BEST_MAP,
                             'best_epoch': BEST_EPOCH
                             }
        apputils.save_checkpoint(epoch, 'RFB_VGG_ML', model, optimizer=optimizer, scheduler=compression_scheduler,
                                 extras=checkpoint_extras, is_best=is_best, name=args.name, dir=msglogger.logdir)
```

--------------------------------------------------------------------------------
The test fn
```
def test(dataset, model, criterion,pylogger, args, epoch, thresh=0.001):
    network = model
    network.eval()
    losses = {'objective_loss': tnt.AverageValueMeter()}
   
    torch.set_grad_enabled(False)
    decoder = Decoder(NUM_CLASSES, mparams, bs=1)
    decoder = decoder.cuda()

    transform = BaseTransform(IMAGE_SIZE, RGB_MEAN, (2, 0, 1))

    num_images = len(dataset)
    print('\nEval on', num_images, 'images. Batch size: 1')

    # store all detections in a dictionary
    # detections_dict['cid'] = {}: dictionary to store all detections for class 'cid'
    # detections_dict['cid']['image_file']: dets for class 'cid' in 'image_file'
    detections_dict = {}
    for cid in range(1, NUM_CLASSES):
        detections_dict[cid] = {}
        for img_index in range(len(dataset)):
            image_file = dataset.image_list[img_index]
            detections_dict[cid][image_file] = np.array([])

    CPU_NMS = common_params['cpu_nms']
    MAX_PER_IMAGE = common_params['mpi']

    total_detection_time = 0.0
    timer = Timer()
    dtlist = []
    
    end = time.time()
    for i in range(num_images):

        #if i%500==0: print('Image', i, '/', num_images, flush=True)

        all_boxes = [[] for _ in range(NUM_CLASSES)]
        img, image_file = dataset.load_image(i)

        scale = torch.Tensor([img.shape[1], img.shape[0], img.shape[1], img.shape[0]])

        x = transform(img).unsqueeze(0)

        timer.tic()

        x = x.cuda()
        scale = scale.cuda()

        # forward pass
        locs, scores = network.forward(x, 'eval')
        # get detections, each box: x1,y1, x2,y2 (normalized to 0-1)
        boxes, scores = decoder.forward(locs, scores)

        # torch.Size([32756, 4])
        boxes = boxes[0]
        # torch.Size([32756, 2])
        scores = scores[0]

        # scale each detection from 0-1 back to image size
        boxes *= scale

        for j in range(1, NUM_CLASSES):
            inds = torch.nonzero(scores[:,j]>thresh).view(-1)

            if len(inds) == 0:
                all_boxes[j] = np.empty([0, 5], dtype=np.float32)
                continue

            cls_boxes = boxes[inds, :]
            cls_scores = scores[inds, j]

            cls_dets = torch.cat((cls_boxes, cls_scores.unsqueeze(1)), 1)

            cls_dets = cls_dets.cpu().numpy()
            keep = nms(cls_dets, 0.50, force_cpu=CPU_NMS)
            cls_dets = cls_dets[keep, :]

            all_boxes[j] = cls_dets

        if MAX_PER_IMAGE > 0:
            image_scores = np.hstack([all_boxes[j][:, -1] for j in range(1, NUM_CLASSES)])
            if len(image_scores) > MAX_PER_IMAGE:
                image_thresh = np.sort(image_scores)[-MAX_PER_IMAGE]
                for j in range(1, NUM_CLASSES):
                    keep = np.where(all_boxes[j][:, -1] >= image_thresh)[0]
                    all_boxes[j] = all_boxes[j][keep, :]

        dtlist.append(int(1000*timer.toc()))
        total_detection_time += timer.toc()

        for j in range(1, NUM_CLASSES):
            detections_dict[j][image_file] = all_boxes[j]
```

------------------------------------------------------------------------------
The train fn:
```
def train(train_loader, model, anchors, criterion, optimizer, epoch, compression_scheduler, loggers, args):
    """
       train_loader = data.DataLoader(trainset, BATCH_SIZE[0], shuffle=True, num_workers=8, collate_fn=detection_collate, drop_last=True)
       criterion = MultiBoxLoss(NUM_CLASSES, overlap_thresh=0.5, negpos_ratio=3)
       anchors = Anchors(mparams)
       optimizer = optim.Adam(params, lr=LR[0], weight_decay=WD[0])
       
       # Statge 2
       train_loader = data.DataLoader(trainset, BATCH_SIZE[1], shuffle=True, num_workers=8, collate_fn=detection_collate, drop_last=True)
    """

    
    torch.set_grad_enabled(True)
  
    
    losses = OrderedDict([(OVERALL_LOSS_KEY, tnt.AverageValueMeter()),
                          (OBJECTIVE_LOSS_KEY, tnt.AverageValueMeter())])

    #classerr = tnt.ClassErrorMeter(accuracy=True, topk=(1, 5))
    batch_time = tnt.AverageValueMeter()
    data_time = tnt.AverageValueMeter()

    # For Early Exit, we define statistics for each exit
    # So exiterrors is analogous to classerr for the non-Early Exit case
    if args.earlyexit_lossweights:
        args.exiterrors = []
        for exitnum in range(args.num_exits):
            args.exiterrors.append(tnt.ClassErrorMeter(accuracy=True, topk=(1, 5)))
    

    import math
    total_samples = len(train_loader.sampler)
    batch_size = train_loader.batch_size
    steps_per_epoch = math.ceil(total_samples / batch_size)
    
    
    model.train()
    #map_stats = []
    end = time.time()
    # Start Traing  
    for train_step, (inputs, targets) in enumerate(train_loader):
        data_time.add(time.time() - end)
       
        inputs = inputs.to(args.device)
        with torch.no_grad():
            targets = [t.to(args.device)for t in targets] # targets is a list but not a tensor
       
       
        # Execute the forward phase, compute the output and measure loss
        if compression_scheduler:
            compression_scheduler.on_minibatch_begin(epoch, train_step, steps_per_epoch, optimizer)
        
        #pdb.set_trace()
        if not hasattr(args, 'kd_policy') or args.kd_policy is None:
            output = model(inputs)
        else:
            output = args.kd_policy.forward(inputs) # A list
            
            
  
        # Please take care here.
        loc, conf = output[:,:, 0:4], output[:,:,4:]
        if not args.earlyexit_lossweights:
            #pdb.set_trace()
            loss_l, loss_c = criterion( (loc, conf), anchors, targets)
            loss = loss_l + loss_c
            # Measure accuracy
            #classerr.add(output.data, target)
         
        else:
            # Measure accuracy and record loss
            loss = earlyexit_loss((loc, conf), target, criterion, args)
        
        # Record loss
        #losses[OBJECTIVE_LOSS_KEY].add(loss.item())
        
        if compression_scheduler:
            # Before running the backward phase, we allow the scheduler to modify the loss
            # (e.g. add regularization loss)
            agg_loss = compression_scheduler.before_backward_pass(epoch, train_step, steps_per_epoch, loss,
                                                                  optimizer=optimizer, return_loss_components=True)
            loss = agg_loss.overall_loss
            losses[OVERALL_LOSS_KEY].add(loss.item())

            for lc in agg_loss.loss_components:
                if lc.name not in losses:
                    losses[lc.name] = tnt.AverageValueMeter()
                losses[lc.name].add(lc.value.item())
        else:
            losses[OVERALL_LOSS_KEY].add(loss.item())
        # STAGE 2 TRAINING: Train the whole network
        # params = filter(lambda p: p.requires_grad, network.module.parameters())
        
        # Compute the gradient and do SGD step
        optimizer.zero_grad()
        loss.backward()
        if compression_scheduler:
            compression_scheduler.before_parameter_optimization(epoch, train_step, steps_per_epoch, optimizer)
        optimizer.step()
        if compression_scheduler:
            compression_scheduler.on_minibatch_end(epoch, train_step, steps_per_epoch, optimizer)
        
        # measure elapsed time
        batch_time.add(time.time() - end)
        steps_completed = (train_step+1)
        
        if steps_completed % args.print_freq == 0:
            # Log some statistics
            errs = OrderedDict()
            if not args.earlyexit_lossweights:
                errs['loss'] = loss.item()
                
            else:
                # for Early Exit case, the Top1 and Top5 stats are computed for each exit.
                for exitnum in range(args.num_exits):
                    errs['loss_exit' + str(exitnum)] =  loss.item()
                    

            stats_dict = OrderedDict()
            for loss_name, meter in losses.items():
                stats_dict[loss_name] = meter.mean
            stats_dict.update(errs)
            stats_dict['LR'] = optimizer.param_groups[0]['lr']
            stats_dict['Time'] = batch_time.mean
            stats = ('Performance/Training/', stats_dict)

            params = model.named_parameters() if args.log_params_histograms else None
            distiller.log_training_progress(stats,
                                            params,
                                            epoch, steps_completed,
                                            steps_per_epoch, args.print_freq,
                                            loggers
                                            )
        end = time.time()

```

Please take care of the `loc, conf = output[:,:, 0:4], output[:,:,4:]`  in train fn.
For the Knowledge distillation, I have checkd the line 123 of the file '.../distiller/distiller/knowledge_distillation.py' ,
![image](https://user-images.githubusercontent.com/37952703/62678208-65508000-b9e3-11e9-829f-9c852a785222.png)
 
It seems that the out should be a tensor. However, in my case, there is a detection model, which would output a tuple of two tensors (locations, confidence). If I don't do any change to adjust it to the Distiller, there would be errors raised. 

Therefore, I insert one line of code to the forward function of both the teacher model and student model , to concatenate the location tensor, of which the shape should be (batch_size, anchor_size, 4) and the confidence tensor, of which the shape would be (batch_size, anchor_size, 2), together (at line 224)
![image](https://user-images.githubusercontent.com/37952703/62678343-d5f79c80-b9e3-11e9-840e-74710a534925.png)

Therefore, I add the `loc, conf = output[:,:, 0:4], output[:,:,4:]` to before the loss computation to within the train fn to separate them.

Please help! Do you need any other information?

**Comment by Technius** [comment]
`KnowledgeDistillationPolicy` only supports classification. For detection, you will have to implement your own version of `KnowledgeDistillationPolicy` using an object detection distillation loss such as the one described in section 3 of [this paper](https://arxiv.org/abs/1805.06361). I've implemented this for SSD at work, but I don't have permission to share it yet.

**Comment by RizhaoCai** [comment]
> `KnowledgeDistillationPolicy` only supports classification. For detection, you will have to implement your own version of `KnowledgeDistillationPolicy` using an object detection distillation loss such as the one described in section 3 of [this paper](https://arxiv.org/abs/1805.06361). I've implemented this for SSD at work, but I don't have permission to share it yet.
> 
> Edit: some more details. In the policy `forward`, you'll need to store the scores and boxes of both the student and teacher so that you can access them in `before_backward_pass`.

Thank you. It seems that what I need to do is to check the source code of `KnowledgeDistiilationPolicy` and implement my own version `KnowledgeDistiilationPolicy`. Then I should change my code like this?
![image](https://user-images.githubusercontent.com/37952703/62749702-ec5b3200-ba8f-11e9-95d6-cebc0f0968d7.png)

Is there anything I should also take care?

By the way, when I was using filters pruning to prune my RFB-NET (VGG-16), I also get the same problem (A pretrain model got 0 mAP after one epoch training). 

My yaml file is like this:
```
version: 1

pruners:
  low_pruner:
    class: L1RankedStructureParameterPruner_AGP
    initial_sparsity : 0.10
    final_sparsity: 0.50
    group_type: Filters
    weights: [base.0.weight,]
              
lr_schedulers:
  pruning_lr:
    class: StepLR
    step_size: 50
    gamma: 0.10

policies:
  - pruner:
      instance_name : low_pruner
    starting_epoch: 0
    ending_epoch: 30
    frequency: 2
```
Should I also implement my own version of 'L1RankedStructureParameterPruner_AGP' ? Particularly do some accommodation to the loss processing?
